### DISCUSSION TITLE
automatic saving of settings on exit should not be the default

### DISCUSSION THREAD

**Discussion by Gene703122** [nvaccess/nvda#13256] [Ideas]
### Is your feature request related to a problem? Please describe.
Settings can be automatically lost or saved when automatic save on exit is used.  This makes changing settings to see what they do more likely to cause unexpected problems.  New users should be free to change settings without having them accidentally be saved on exit or shut down or restart of Windows.  
### Describe the solution you'd like
The default should be that savings be manually saved with control+NVDA+c.  If it is worried that new users won't know how to save settings, when NVDA closes down a do you want to save settings with a don't show me this again dialog can be provided.
There are no alternatives I can see.
### Additional context
I, an experienced user have twice lost all my settings.  I reverted to factory defaults to check something, shut down NVDA and lost all my settings because of the default behavior.

**Comment by feerrenrut** [comment]
I don't think it's likely that the default setting for NVDA will be changed, it isn't as user friendly in the common case.

An alternative might be to make exporting / importing settings easier. IE a manual option in tools. Then make sure a backup of settings is made when reverting to defaults.

**Comment by Gene703122** [comment]
Expecting people to manually issue a command when they want to save settings, is hardly a burden on users.  There is such a thing as too much automation.  

Window-eyes, Openbook, and Kurzweil all require an affirmative command to permanently save settings.  In all these programs, it is common that users change settings with no intention of saving them.
      
It is too easy to have settings accidentally be saved.  A spontaneous computer reboot will result in settings being saved when not intended.  Unloading NVDA to use another screen-reader will unintentionally save settings.  Shutting down or restarting Windows will unintentionally save settings and users aren't necessarily going to remember to restore the previous configuration before doing these things.  
      
Word processors and some other programs routinely ask users if they want to save changes if they haven't manually done so before exiting.  
Good safety practices shouldn't be sacrificed on the altar of convenience and we are talking about one command.  
      
Two alternatives I didn't think of before present themselves.  Have NVdA ask, as a word processor does, if you want to save changes when shutting down.  Or every time the current configuration is saved, the previous configuration should be saved as a backup and should be able to be restored by issuing one command.  Or both might be implemented.  
      
I think being able to export and import settings as well as speech dictionaries and the pronunciation settings and punctuation/symbol pronunciation settings is important and should be implemented but I don't consider that to be a good solution to this problem.  It is more complex than issuing one affirmative command.

**Comment by josephsl** [comment]
Hi, one thing we need to consider: suppose a prompt is implemented to let users save settings on exit. What would happen if people change settings from multiple configuration profiles? Note that save settings option is global across all profiles, so that should be taken into account. Thanks.

**Comment by CyrilleB79** [comment]
> Hi, one thing we need to consider: suppose a prompt is implemented to let users save settings on exit. What would happen if people change settings from multiple configuration profiles? Note that save settings option is global across all profiles, so that should be taken into account. Thanks.

Actually, the fact that the "Save settings when exiting NVDA" option is global and controls all the profiles is an issue. It is discussed in #10156 comments.

Also to note:
Taking the example of Word again: If you have modified the current document, you will be prompted to save or discard your changes. But if you change one of the options or the content of the ribbons, you wont't be asked any confirmation when exiting the program.
The issue with NVDA options is that some options may be changed very easily with shortcut keys; e.g. speech speed. Other ones are much less commonly changed e.g. report tabs with sound or speech. A user may expect that the one less commonly used be saved each time it is modified whereas the one changed more frequently may not be expected to be saved each time.

**Comment by Gene703122** [comment]
It looks to me as though people are taking something simple and making it complicated.  Having different saving behaviors for different settings is unreasonably complicated.  Its micromanaging behavior, resulting in unreasonably complicated behavior.

I am trying to accomplish only one thing, avoiding accidental and unwanted saving of settings.  I originally said that automatic saving shouldn't be the default.  As a result of further thought because of the discussion, I now think a dialog should be used.  I think this behavior should be uniform.  Whenever a setting is changed anywhere, in a profile or the main program, the same dialog would come up asking if the changes should be saved.  

As I understand matters, this is consistent with how NVDA saves settings.  At present, any change anywhere is automatically saved on exit.  All I'm doing is having a dialog come up during exit.  

also, a dialog does away with another objection.  If someone now sets NVDA not to automatically save settings, they may loose settings changes they want to keep by not manually saving them.  A dialog eliminates both accidental saving or accidental not saving settings.

**Comment by CyrilleB79** [comment]
> It looks to me as though people are taking something simple and making it complicated.

Maybe because the issue is simple but the way to solve it isn't.

> Having different saving behaviors for different settings is unreasonably complicated. Its micromanaging behavior, resulting in unreasonably complicated behavior.

To clarify my previous comment, I do not advocate for splitting the options in two sets according to how often they are modified. I was just saying that the wide variety of options makes it difficult to find a good solution to this issue.


> I am trying to accomplish only one thing, avoiding accidental and unwanted saving of settings. I originally said that automatic saving shouldn't be the default. As a result of further thought because of the discussion, I now think a dialog should be used. I think this behavior should be uniform. Whenever a setting is changed anywhere, in a profile or the main program, the same dialog would come up asking if the changes should be saved.
> As I understand matters, this is consistent with how NVDA saves settings. At present, any change anywhere is automatically saved on exit. All I'm doing is having a dialog come up during exit.

I know that NVDA's actual behavior is to handle general settings and profiles the same way. But I really think that there is a real-life use case where you should be able to:
* save general settings on exit
* do not save profile settings on exit to avoid adding more and more options in the profile's config.

One other point also to note: I think that Jaws does not prompt for any confirmation when exiting if some option has been changed. Just mentionning to say that it is an example where the options are saved automatically and used by many users. I do not advocate for following systematically all what Jaws does.

As a last point, in the event where we put a dialog to ask for confirmation when NVDA exits, a question remains unanswered: What happens if NVDA is running and you press ctrl+alt+N? There are two paths:
* The new instance of NVDA is delayed until the "Save modifications" dialog is answered
or
* The old instance of NVDA is killed abruptly and the new one is started immediately

We should consider the control+alt+press in two distinct scenarios:
* Restarting when NVDA is running normally
* Restarting when NVDA is frozen

**Comment by Gene703122** [comment]
As I understand the discussion linked to about saving profile settings, the issue is that the person wants to be able to change profile settings and not have them used if he wants to use the general setting in the  profile while retaining the saved setting.  All sorts of proposals are discussed to allow dealing with this wish.  I don't think that issue should intrude on this one. If a person changes something if new proposals are adopted then, whatever the change, it is going to change something evidently permanently.  You will be changing a setting and only applying it to the ;profile or it will affect general settings and the profile will use those settings.  But what does any of this have to do with my proposal?  Any change of setting that will be saved of any kind, limited or global, would generate the dialog.  I'm not trying to change anything else but to eliminate accidental or unwanted savings of changes.

There may be something I don't understand but I don't see why my proposal causes any problems for what is being discussed in the other topic.

**Comment by Gene703122** [comment]
JAWS doesn't prompt on exit because settings are saved when you activate the OK button in the settings dialog.  That means that whether someone is changing a setting for temporary use or permanent use, it is permanently changed and the only way to revert is to manually change the setting again. 

I consider this to be a bad system.  It means that there are no temporary changes possible.  NVDA allows for temporary changes. So did Window-eyes. 

I'm not sure my proposal has anything to do with the discussion linked to.  but if it does, it seems to me that what I've said would cause the following behavior to be adopted when the dialog comes up.  Any permanent change anywhere would generate the dialog.  Any temporary change in a profile, meant to be discarded, would not cause the dialog to come up. 

As far as the control alt n instance is concerned, it is my understanding that control alt n is intended to be used if NVDA freezes to unload the frozen copy and open the program again.  It seems to me that, since that is its purpose and any time a program freezes, work is lost, that the question doesn't arise.  There is no reason this dialog should function differently than Windows or programs do in general.  When a program crashes, you lose your work.

Gene

**Comment by feerrenrut** [comment]
I'm converting this issue into a discussion. It isn't currently a cohesive description of a problem. I welcome you to continue to explore ideas there.

**Comment by feerrenrut** [comment]
@Gene703122 I think there it is a leap to go from "I didn't mean to save after resetting to defaults", to "users shouldn't have their config persisted in the general case". Most modern software (including Windows) doesn't require you to explicitly save settings. I think we should focus on the specific use case:

"As a user I want to be able to temporarily restore factory defaults, so that I can test something. I don't want to risk losing my config"

I'd like to know more about what "something" is, I suspect that will illuminate another use case.

**Comment by CyrilleB79** [comment]
Quickly checking to Jaws, it is more complicated that what I was writting yesterday.
They have a Setting center for permanent changes and Quick settings allowing to say, for each option, if the change is permanent, until the next restart or until the next app change (alt+tab). I have not digged further but it seems quite complicated and does not seem an example to follow.
